src/actions.rs

```diff
@@ -0,0 +1,886 @@
+    Key(KeyActionChannel),
+    /// A channel containing pointer actions. All actions in the channel are for a single
+    /// pointer type.
+    Pointer(PointerActionChannel),
+}
+
+impl ActionChannel {
+    /// Add a pause action for this channel.
+    pub fn add_pause(&mut self, duration: Duration) {
+        match self {
+            ActionChannel::Null(channel) => channel.add_pause(duration),
+            ActionChannel::Key(channel) => channel.add_pause(duration),
+            ActionChannel::Pointer(channel) => channel.add_pause(duration),
+        }
+    }
+
+    fn into_sequence(self) -> WDActions::ActionSequence {
+        match self {
+            ActionChannel::Null(channel) => WDActions::ActionSequence {
+                id: channel.id,
+                actions: WDActions::ActionsType::Null {
+                    actions: channel.actions.into_iter().map(|x| x.into()).collect(),
+                },
+            },
+            ActionChannel::Key(channel) => WDActions::ActionSequence {
+                id: channel.id,
+                actions: WDActions::ActionsType::Key {
+                    actions: channel.actions.into_iter().map(|x| x.into()).collect(),
+                },
+            },
+            ActionChannel::Pointer(channel) => WDActions::ActionSequence {
+                id: channel.id,
+                actions: WDActions::ActionsType::Pointer {
+                    parameters: WDActions::PointerActionParameters {
+                        pointer_type: channel.pointer_type.into(),
+                    },
+                    actions: channel.actions.into_iter().map(|x| x.into()).collect(),
+                },
+            },
+        }
+    }
+}
+
+/// An Action Chain is simply a list of channels. See the documentation for `ActionChannel`
+/// for more details.
+///
+/// Also see `ActionChainBuilder` for a convenient, high-level way to create an `ActionChain`.
+#[derive(Debug, Clone, Default)]
+pub struct ActionChain {
+    channels: Vec<ActionChannel>,
+}
+
+impl ActionChain {
+    /// Create a new ActionChain.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Create an ActionChainBuilder.
+    pub fn builder() -> ActionChainBuilder {
+        ActionChainBuilder::default()
+    }
+
+    /// Add a complete channel to this ActionChain. This allows more flexibility if you want
+    /// to create your own channels and then add them here.
+    pub fn add_channel(&mut self, channel: ActionChannel) {
+        self.channels.push(channel);
+    }
+}
+
+impl From<ActionChain> for webdriver::command::ActionsParameters {
+    fn from(ac: ActionChain) -> Self {
+        let mut sequences = Vec::new();
+
+        for channel in ac.channels {
+            sequences.push(channel.into_sequence());
+        }
+
+        webdriver::command::ActionsParameters { actions: sequences }
+    }
+}
+
+/// Builder for an ActionChain. Use `ActionChain::builder()` to create one.
+#[derive(Debug)]
+pub struct ActionChainBuilder {
```

Comment:
This feels like something that should live in `thirtyfour`, and not `fantoccini`.

Reply:
I'm open to that. See the above discussion in relation to `Client::perform_actions()` though.

Reply:
Removed.